### DISCUSSION TITLE
Some distro specific packages.

### DISCUSSION THREAD

**Discussion by Masgalor** [lldap/lldap#607] [Show and tell]
I really like your software, but not so much the need to install a container engine.
So I've created some build-recipes at [build.opensuse.org](https://build.opensuse.org) to get distro specific builds/packages.

SUSE is a big fan of reproducible builds, so everything has to be preloaded and there is no internet during buildtime.
This makes packaging rust based projects, where almost anything is downloaded during build not easier...
~So in the end I could not come up a way to build the webapp (for now), because wasm-pack/wasm-bindgen need to download some dependencies and could not be convinced otherwise.~

~Because of that I´ve split the backend (server) and frontend (app) into two packages.
The server is completely built form source, the app only repacks the files from the official release.
In the end you will need to install both, but i think it is more transparent that way.~

Thanks to the help of @nitnelave and @martadinata666 the backend and frontend are now built and packaged together!

The build-project can be found here: [LLDAP Project](https://build.opensuse.org/project/show/home:Masgalor:LLDAP).
The ready-to-install packages can be found here: [Packages/Repositories](https://software.opensuse.org//download.html?project=home%3AMasgalor%3ALLDAP&package=lldap)

At the moment following distributions are supported:

- SUSE Tumbleweed
- SUSE Leap 15.5
- SUSE Leap 15.4
- CentOS 8
- CentOS 8-Stream
- CentOS 9-Stream
- Debian 11
- Fedora 37
- Fedora 38
- Fedora Rawhide
- Ubuntu 18.04
- Ubuntu 20.04
- Ubuntu 22.04
- Ubuntu 23.04

I can add support for more distributions if someone is interested just let me know.
Sidenote: Debian builds are for now limited to `bullseye` because of the recent release of `bookworm`.

**Comment by nitnelave** [comment]
Thanks a lot for packaging it! I'm no expert in packaging, so I'm glad for the help.

For the reproducible build, I suppose you've had a look at passing `--offline` to `cargo build` / `wasm-pack` ? See https://rustwasm.github.io/docs/wasm-pack/commands/build.html#extra-options

**Reply by Masgalor** [reply]
Yes I know.
The problem is not building the app itself `wasm-pack build --target web --release -- --offline` works just fine.
But after completing the build wasm-pack tries to build some (extensions?) for wasm-bindgen and that fails because it tries to use crates.io for this and I dont see a way to change it.

```
[   85s]    Compiling lldap_app v0.4.3 (/home/abuild/rpmbuild/BUILD/app)
[  128s]     Finished release [optimized] target(s) in 1m 11s
[  128s] [WARN]: :-) origin crate has no README
[  128s] [INFO]: Installing wasm-bindgen...
[  133s]     Updating crates.io index
[  133s] warning: spurious network error (2 tries remaining): failed to resolve address for github.com: Temporary failure in name resolution; class=Net (12)
[  133s] warning: spurious network error (1 tries remaining): failed to resolve address for github.com: Temporary failure in name resolution; class=Net (12)
[  133s] error: could not find `wasm-bindgen-cli` in registry `crates-io` with version `=0.2.84`
[  133s] Error: Installing wasm-bindgen with cargo
[  133s] Caused by: Installing wasm-bindgen with cargo
[  133s] Caused by: failed to execute `cargo install`: exited with exit status: 101
```

**Reply by nitnelave** [reply]
oh, I see. This might be something you can raise at https://github.com/rustwasm/wasm-pack

Otherwise, you might want to look into https://crates.io/crates/cargo-local-registry ? It seems it can help with avoiding crates.io accesses.

**Reply by nitnelave** [reply]
More info here: https://doc.rust-lang.org/cargo/reference/source-replacement.html

**Reply by Masgalor** [reply]
I will definitely keep my eyes open for a solution.

Thank you for your feedback!

**Reply by martadinata666** [reply]
Trying to replicate this, I can't say that im fully offline as I still download cargo deps.
```
   Compiling http v0.2.9
   Compiling yew-router v0.16.0
   Compiling graphql_client v0.10.0
   Compiling yew_form_derive v0.1.8 (https://github.com/jfbilodeau/yew_form?rev=4b9fabffb63393ec7626a4477fd36de12a07fac9#4b9fabff)
   Compiling lldap_auth v0.3.0 (/home/debian/lldap/auth)
   Compiling jwt v0.13.0
   Compiling url-escape v0.1.1
   Compiling lldap_app v0.5.0-alpha (/home/debian/lldap/app)
    Finished release [optimized] target(s) in 3m 53s
[INFO]: ⬇️  Installing wasm-bindgen...
Error: Not able to find or install a local wasm-bindgen.
Caused by: Not able to find or install a local wasm-bindgen.
```

My suggested workaround, is using prebuilt, place the binary to `/usr/local/bin/` and `Cargo.toml` need to modified to exact version of `wasm-bindgen`, `0.2` to `0.2.87`, else it wont be detected correctly.
Command I'm using: `wasm-pack build --target web --release --mode no-install -- --offline`

Local bin:
```
debian@89bbf2e962d6:~/lldap/app$ ls ~/.local/bin/
wasm2es6js  wasm-bindgen  wasm-bindgen-test-runner  wasm-pack
```

Build log:
```
debian@89bbf2e962d6:~/lldap/app$ wasm-pack build --target web --release --mode no-install -- --offline
[INFO]: 🎯  Checking for the Wasm target...
info: downloading component 'rust-std' for 'wasm32-unknown-unknown'
info: installing component 'rust-std' for 'wasm32-unknown-unknown'
 17.9 MiB /  17.9 MiB (100 %)   5.4 MiB/s in  3s ETA:  0s
[INFO]: 🌀  Compiling to Wasm...
    Finished release [optimized] target(s) in 1.70s
[INFO]: Skipping wasm-opt as no downloading was requested
[INFO]: License key is set in Cargo.toml but no LICENSE file(s) were found; Please add the LICENSE file(s) to your project directory
[INFO]: ✨   Done in 18.88s
[INFO]: 📦   Your wasm pkg is ready to publish at /home/debian/lldap/app/pkg.
```
It still downloading, rust target `wasm32-unknown-unknown` but you maybe already handling this ☕ 


Use pre-builts
https://github.com/rustwasm/wasm-pack/
https://github.com/rustwasm/wasm-bindgen/

Additional Refs
* https://github.com/rustwasm/wasm-pack/issues/308

**Reply by martadinata666** [reply]
~~Maybe our CI need to use this also, as this I assume prevents building the applications? 🤔~~ My bad, I had target cache mounted in 😮‍💨 I thought we can get away only building the `pkg`

**Reply by nitnelave** [reply]
Our CI doesn't try to build offline, so it doesn't break it to have to fetch packages. (Actually our CI is not truly reproducible, as it gets new versions of rust, with sometimes clippy adding new warnings which break the checks)

**Reply by martadinata666** [reply]
So for additional info for this disscussion, as `wasm-pack` somehow hiding useful logs. This is a log, with `RUST_LOG=debug`

```
cd app
  wasm-pack build --target web --release --mode no-install -- --offline
  shell: sh -e {0}
  env:
    CARGO_TERM_COLOR: always
    RUST_LOG: debug
[INFO  wasm_pack::command] Running build command...
[DEBUG ureq::stream] connecting to crates.io:443 at 18.155.202.15:443
[DEBUG rustls::client::hs] No cached session for DnsName("crates.io")
[DEBUG rustls::client::hs] Not resuming any session
[DEBUG rustls::client::hs] Using ciphersuite TLS13_AES_128_GCM_SHA256
[DEBUG rustls::client::tls13] Not resuming
[DEBUG rustls::client::tls13] TLS1.3 encrypted extensions: [ServerNameAck]
[DEBUG rustls::client::hs] ALPN protocol is None
[DEBUG ureq::stream] created stream: Stream(RustlsStream)
[DEBUG ureq::unit] sending request GET https://crates.io/api/v1/crates/wasm-pack
[DEBUG ureq::unit] writing prelude: GET /api/v1/crates/wasm-pack HTTP/1.1
    Host: crates.io
    User-Agent: wasm-pack/0.12.0 (https://github.com/rustwasm/wasm-pack)
    Accept: */*
    accept-encoding: gzip
[DEBUG ureq::response] Chunked body in response
[DEBUG ureq::unit] response 200 to GET https://crates.io/api/v1/crates/wasm-pack
[DEBUG ureq::pool] adding stream to pool: https|crates.io|443 -> Stream(RustlsStream)
[DEBUG ureq::stream] dropping stream: Stream(RustlsStream)
[INFO  wasm_pack::command::build] Checking rustc version...
[INFO  wasm_pack::command::build] rustc version is 70.
[INFO  wasm_pack::command::build] Checking crate configuration...
[INFO  wasm_pack::command::build] Crate is correctly configured.
[INFO  wasm_pack::command::build] Checking for wasm-target...
[INFO]: 🎯  Checking for the Wasm target...
[INFO  wasm_pack::build::wasm_target] Looking for wasm32-unknown-unknown in "/home/debian/rustup/toolchains/1.70.0-x86_64-unknown-linux-gnu/lib/rustlib"
[INFO  wasm_pack::build::wasm_target] Found wasm32-unknown-unknown in "/home/debian/rustup/toolchains/1.70.0-x86_64-unknown-linux-gnu/lib/rustlib"
[INFO  wasm_pack::command::build] Checking for wasm-target was successful.
[INFO  wasm_pack::command::build] Building wasm...
[INFO]: 🌀  Compiling to Wasm...
[INFO  wasm_pack::child] Running cd "/__w/lldap/lldap/app" && "cargo" "build" "--lib" "--release" "--target" "wasm32-unknown-unknown" "--offline"
   Compiling lldap_auth v0.3.0 (/__w/lldap/lldap/auth)
   Compiling lldap_app v0.5.0-alpha (/__w/lldap/lldap/app)
    Finished release [optimized] target(s) in 35.27s
[INFO  wasm_pack::command::build] wasm built at "/__w/lldap/lldap/app/target/wasm32-unknown-unknown/release".
[INFO  wasm_pack::command::build] Creating a pkg directory...
[INFO  wasm_pack::command::build] Created a pkg directory at "/__w/lldap/lldap/app".
[INFO  wasm_pack::command::build] Identifying wasm-bindgen dependency...
[INFO  wasm_pack::command::build] Installing wasm-bindgen-cli...
[INFO]: ⬇️  Installing wasm-bindgen...
[WARN  wasm_pack::install] could not download pre-built `wasm-bindgen`: wasm-bindgen v0.2.84 is not installed!. Falling back to `cargo install`.
[DEBUG wasm_pack::install] Attempting to use a `cargo install`ed version of `wasm-bindgen=0.2.84`
[INFO  wasm_pack::command::build] Installing wasm-bindgen-cli was successful.
[INFO  wasm_pack::command::build] Building the wasm bindings...
Error: Not able to find or install a local wasm-bindgen.
Caused by: Not able to find or install a local wasm-bindgen.
Error: Process completed with exit code 1.
```

**Reply by Masgalor** [reply]
Thank you very much for your input, I am rather used to comments like `meh, who cares, we have docker`.

So I went to `./lldap/app/Cargo.toml` and changed `wasm-bindgen = "0.2" to wasm-bindgen = "0.2.84"`.
Then I went to `https://github.com/rustwasm/wasm-bindgen/releases` and downloaded the precompiled binaries for version `0.2.84`.
I placed them next to the binary `wasm-pack` and that indeed leads to a promising result, there is still a warning but the build-process finishes without error.

```
[   44s] + wasm-pack build --target web --release --mode no-install -- --offline
...
[  148s]    Compiling lldap_app v0.4.3 (/home/abuild/rpmbuild/BUILD/app)
[  192s]     Finished release [optimized] target(s) in 2m 26s
[  192s] [WARN]: :-) origin crate has no README
[  192s] [INFO]: Skipping wasm-opt as no downloading was requested
[  192s] [INFO]: Optional fields missing from Cargo.toml: 'description', 'repository', and 'license'. These are not necessary, but recommended
[  192s] [INFO]: :-) Done in 2m 26s
[  192s] [INFO]: :-) Your wasm pkg is ready to publish at /home/abuild/rpmbuild/BUILD/app/pkg.
[  192s] [WARN]: :-) [6] Couldn't resolve host name (Could not resolve host: crates.io)
```

I will see if the resulting files are fine and think about how to manage this.
Preloading binaries is fine, but it seems all of them need to match the exact version of each other, thats kind of cumbersome.

**Reply by martadinata666** [reply]
I saw the logs, I dunno what it mean for the half of it 🤣, but essentially on frontend, I can say that you just copying things and add some files also with this structure.
![Cuplikan layar dari 2023-06-19 22-06-53](https://github.com/lldap/lldap/assets/2433562/eda2cbff-e241-4ce8-850a-4bc6f380715f)


The additional files in this path:
* app/static/libraries.txt
* app/static/fonts/fonts.txt

**Reply by Masgalor** [reply]
Yeah `lldap-frontend` just grabs the latest release on github, extracts the app folder and puts it in the right place. :sweat_smile:
The package `lldap-frontend-dev` builds the frontend from source, after that works I will put everything in the main lldap package. :wink:

**Reply by nitnelave** [reply]
The exact version of `wasm-bindgen` should be in `Cargo.lock`. Otherwise, I'm fine with checking in a specific version in `Cargo.toml`, but I think using `Cargo.lock` is a better approach.

**Reply by Masgalor** [reply]
I've finished the final build-recipes to compile the server and the app in the same run and package them together.

It is not necessary to change `Cargo.toml`, the important part is to preload the compiled version of `wasm-bindgen` that matches the version in `Cargo.lock` and put them in the right place for `wasm-pack`.

I still need a better logic to bundle the fonts and statics, but there is always something to improve.
For now I am happy that the server and app are built and packed together.

**Reply by nitnelave** [reply]
If it helps, the Dockerfile just does a wget for the static resources: https://github.com/lldap/lldap/blob/main/Dockerfile#L86-L87

**Reply by martadinata666** [reply]
additional info for the static part, if you are using `index.html` the static resource will fetch from internet, and `index_local.html` will fetch on the local disk. There is previous issue that the web borked because the server that hosted lldap lost internet connection.

**Comment by nitnelave** [comment]
Also, you might be interested in https://github.com/lldap/lldap/issues/579
